### PR TITLE
feat(provider/dcos): Server group and instance caching agents

### PR DESCRIPTION
**Pull request by willgorman**
This pull request adds the instance and server group caching agents for DC/OS along with some related classes and tests.

### CODE REVIEW

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/mapper/AppToDeployDcosServerGroupDescriptionMapper.groovy [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    desc.constraints = app.constraints?.stream()?.map({ constraintParts -> constraintParts.join(':') })?.collect(joining(','))
+
+    desc.fetch = app.fetch?.collect({ f ->
+      new DeployDcosServerGroupDescription.Fetchable().with {
+        uri = f.uri
+        cache = f.cache
+        extract = f.extract
+        executable = f.executable
+        outputFile = f.outputFile
+        it
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/deploy/util/mapper/AppToDeployDcosServerGroupDescriptionMapper.groovy]: Please use `return` statements for all of these.

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/DcosProviderUtils.groovy [hunk 2]
```diff
@@ -0,0 +1,169 @@
...
+    } else {
+      result.tag = image.substring(cIndex + 1)
+      result.repository = image.subSequence(0, cIndex)
+    }
+
+    normalizeImageDescription(result)
+    result
+  }
+
+  static Void normalizeImageDescription(ImageDescription image) {
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/DcosProviderUtils.groovy]: Any reason for this to be `Void` wrapper-type?

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/MutableCacheData.groovy [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+ */
+
+package com.netflix.spinnaker.clouddriver.dcos.provider
+
+import com.fasterxml.jackson.annotation.JsonCreator
+import com.fasterxml.jackson.annotation.JsonProperty
+import com.netflix.spinnaker.cats.cache.CacheData
+
+/* TODO(lwander) this was taken from the netflix cluster caching, and should probably be shared between all providers. */
+// I'm sorry, I copied this again (wgorman)
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/MutableCacheData.groovy]: heh, welcome to the club

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/DcosProviderUtils.groovy [hunk 4]
```diff
@@ -0,0 +1,169 @@
...
+    if (!image.registry) {
+      image.registry = "hub.docker.com" // TODO configure or pull from docker registry account
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/DcosProviderUtils.groovy]: Would this be less hidden as a default value of the `ImageDescription.registry` field itself?

File: clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosInstanceCachingAgent.groovy [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+  CacheResult loadData(ProviderCache providerCache) {
+    log.info("Loading tasks in $agentType")
+
+    // The tasks API returns all tasks, but we want to ensure we only cache ones valid for the current account.
+    def tasks = dcosClient.getTasks().tasks.findAll {
+      DcosSpinnakerAppId.parse(it.appId, accountName).isPresent()
+    }
+    def deployments = dcosClient.getDeployments()
+
+    buildCacheResult(tasks, deployments)
```
Comment (duftler) [inline clouddriver-dcos/src/main/groovy/com/netflix/spinnaker/clouddriver/dcos/provider/agent/DcosInstanceCachingAgent.groovy]: Please use `return` statement here and below.

**Comment by willgorman** [comment]
Latest commit should address all comments.

**Pull request merged**